packages/core/contracts/financial-templates/funding-rate-store/implementation/OptimisticOracle.sol

```diff
@@ -0,0 +1,357 @@
+    }
+
+    function proposePriceFor(
+        address proposer,
+        address requester,
+        bytes32 identifier,
+        uint256 timestamp,
+        int256 proposedPrice
+    ) public nonReentrant() {
+        Request storage request = _getRequest(requester, identifier, timestamp);
+        require(getState(requester, identifier, timestamp) == State.Requested, "proposePriceFor: Requested");
+        request.proposer = proposer;
+        request.proposedPrice = proposedPrice;
+        // If a custom liveness has been set
+        request.expirationTime = getCurrentTime().add(
+            request.customLiveness != 0 ? request.customLiveness : defaultLiveness
+        );
+        request.currency.safeTransferFrom(msg.sender, address(this), request.bond.add(request.finalFee));
+
+        // Event.
+        emit ProposePrice(requester, proposer, identifier, timestamp, proposedPrice);
+
+        // Callback.
+        try FundingRateRequester(requester).priceProposed(identifier, timestamp) {} catch {}
+    }
+
+    function proposePrice(
+        address requester,
+        bytes32 identifier,
+        uint256 timestamp,
+        int256 proposedPrice
+    ) external {
+        // Note: re-entrancy guard is done in the inner call.
+        proposePriceFor(msg.sender, requester, identifier, timestamp, proposedPrice);
+    }
+
+    function disputePriceFor(
+        address disputer,
+        address requester,
+        bytes32 identifier,
+        uint256 timestamp
+    ) public nonReentrant() {
+        Request storage request = _getRequest(requester, identifier, timestamp);
+        require(getState(requester, identifier, timestamp) == State.Proposed, "disputePriceFor: Proposed");
+        request.disputer = disputer;
+        request.currency.safeTransferFrom(msg.sender, address(this), request.bond.add(request.finalFee));
+        _getOracle().requestPrice(identifier, timestamp);
+        StoreInterface store = _getStore();
+        request.currency.safeIncreaseAllowance(address(store), request.finalFee);
+        _getStore().payOracleFeesErc20(address(request.currency), FixedPoint.Unsigned(request.finalFee));
+
+        // Compute refund.
+        uint256 refund = 0;
+        if (request.reward != 0 && request.refundOnDispute) {
+            refund = request.reward;
+            request.reward = 0;
+        }
+
+        // Event.
+        emit DisputePrice(requester, request.proposer, disputer, identifier, timestamp);
+
+        // Callback.
+        try FundingRateRequester(requester).priceDisputed(identifier, timestamp, refund) {} catch {}
+    }
+
+    function disputePrice(
+        address requester,
+        bytes32 identifier,
+        uint256 timestamp
+    ) external {
+        // Note: re-entrancy guard is done in the inner call.
+        disputePriceFor(msg.sender, requester, identifier, timestamp);
+    }
+
+    function getPrice(bytes32 identifier, uint256 timestamp) external nonReentrant() returns (int256) {
```

Comment:
this method will revert if the State is not `Expired` or `Resolved` due to the `_settle` function. This is similar behavior to the `Voting` contract's `GetPrice` method. 

Do you want to add a `hasPrice` function to enable a non-reverting function to check if the `OptimisticOracle` has a price?

Reply:
with that said, the right answer could be to simply call `getState` and validate if the request is within a state that will return a price & not revert when calling `getPrice`.

Reply:
Not super opinionated here. Was trying to reduce the number of methods offered initially, but ultimately ended up adding `getState` and `getRequest` to provide users with more info where needed. I was initially thinking you would just call `getPrice` inside a try-catch instead of using `hasPrice`, but would be okay adding a `hasPrice` too. Thoughts?

Reply:
A `hasPrice`, as opposed to a `getState`, reduces burden on the calling contract but trades off with less information. It would also match the Voting.sol interface which might be useful